From 817a44969cc629dc25805b2a58c9d25fc8baeb15 Mon Sep 17 00:00:00 2001
From: "Prakash, Divya1" <divya1.prakash@intel.com>
Date: Thu, 14 Sep 2017 22:02:01 +0530
Subject: [PATCH 0889/2367] ASoC: Intel: Skylake: Reset DSP pipe in
 skl_pcm_hw_free

Currently during destroy pipeline the gateway is disabled
before DMA completion. This leads to improper draining of
data and subsequently causing issues on HD-Audio DMA.
Hence added a new pipe reset IPC in skl_pcm_hw_free in
which the Gateway Enable(GEN bit) is reset to 0 after
DMA completion in skl_pcm_trigger.

Change-Id: I4786dea60428e2918df139fb4ffc196f8d8ef6dc
Signed-off-by: Prakash, Divya1 <divya1.prakash@intel.com>
Reviewed-on: https://git-gar-1.devtools.intel.com/gerrit/19217
Reviewed-by: Koul, Vinod <vinod.koul@intel.com>
Tested-by: Sm, Bhadur A <bhadur.a.sm@intel.com>
---
 sound/soc/intel/skylake/skl-pcm.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/sound/soc/intel/skylake/skl-pcm.c b/sound/soc/intel/skylake/skl-pcm.c
index 61fb43b..e623d55 100644
--- a/sound/soc/intel/skylake/skl-pcm.c
+++ b/sound/soc/intel/skylake/skl-pcm.c
@@ -429,13 +429,23 @@ static int skl_pcm_hw_free(struct snd_pcm_substream *substream,
 	struct hdac_ext_stream *stream = get_hdac_ext_stream(substream);
 	struct hdac_stream *hstream = hdac_stream(stream);
 	struct snd_pcm_runtime *runtime = substream->runtime;
+	struct skl *skl = get_skl_ctx(dai->dev);
+	struct skl_module_cfg *mconfig;
+	int ret;
 
 	dev_dbg(dai->dev, "%s: %s\n", __func__, dai->name);
 
+	mconfig = skl_tplg_fe_get_cpr_module(dai, substream->stream);
+
 	if (runtime->no_rewinds) {
 		snd_hdac_ext_stream_set_spib(ebus, stream, 0);
 		snd_hdac_ext_stream_spbcap_enable(ebus, 0, hstream->index);
 	}
+	if (mconfig) {
+		ret = skl_reset_pipe(skl->skl_sst, mconfig->pipe);
+		if (ret < 0)
+			dev_err(dai->dev, "%s:Reset failed ret =%d", __func__, ret);
+	}
 
 	snd_hdac_stream_cleanup(hdac_stream(stream));
 	hdac_stream(stream)->prepared = 0;
-- 
2.7.4

